OCPBUGS-65792: [release-4.19] e2e gcp ocl PR backport

umohnani8 opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**
This is a manual backport of https://github.com/openshift/machine-config-operator/pull/5268 and https://github.com/openshift/machine-config-operator/pull/5171 to make the e2e-ocl tests stable.

A similar backport for 4.20 was done in https://github.com/openshift/machine-config-operator/pull/5407.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

cheesesashimi committed:
adds ImagePruner and ImageRegistrySecret

This moves all of the image pruning implementation into a separate
struct and enforces a separation of concerns between the reconciler and
the image pruning tools. Additionally, this also allows the injection of
a fake imagepruner into the build controller for unit testing purposes.

This also introduces a centralized image registry secret wrapper that
can translate image registry secrets between multiple formats. This
allows more streamlined consumption and merging of image registry secrets.

cheesesashimi committed:
adds secretconverter devex helper

cheesesashimi committed:
remove empty keys from PolicySignature JSON blob

In containers/image v5.35.0, several fields under
signature.PolicyRequirement had the `omitempty` JSON tag added to their
definitions. Both TestGeneratePolicyJSON and
TestUpdateNamespacedPolicyJSONs failed because their expected data
assumed those fields would be there, even if they were empty.

cheesesashimi committed:
detect if image exists or access is denied

This also adds e2e tests for edge-cases that may have been missed as
well as a test-case that exposes the clusters' internal image registry
(if it exists) and validates the error handling functions against it.

The exposure functions were repackaged from the devex helpers for easy
reuse across both contexts.

cheesesashimi committed:
check whether the image exists before trying to reuse it for MachineConfigPools

cheesesashimi committed:
e2e test should check for init container status instead of overall pod status

cheesesashimi committed:
move kernel type test into rollout test

cheesesashimi committed:
install skopeo from source, when missing

cheesesashimi committed:
speed up e2e-ocl by deleting build pods

This also modifies the MachineConfigPool status updates to use the
constants we've added for that purpose instead of non-authoritative
strings.

cheesesashimi committed:
adds additional checks to annotation test

cheesesashimi committed:
consider UID when determining MOSB existence

The Assert() helper only considers the MOSB name and not its UID. This
causes a problem where the MOSB is replaced, but the assertion times out
waiting for the replacement since it does not know that it was actually
replaced.

umohnani8 commented:
/test e2e-gcp-op-ocl

umohnani8 committed:
Fix backport compatibility issues for 4.19

This commit addresses several compatibility issues when backporting
PRs #5171 and #5407 to the 4.19 branch:

1. Remove API changes:
   - Eliminated all BuildDegraded condition code to avoid requiring the
     MachineConfigPoolImageBuildDegraded constant that doesn't exist in
     4.19 API

2. Library compatibility for containers/image v5.34.1:
   - Simplified IsAccessDeniedErr to work without
     docker.UnexpectedHTTPStatusError (not available in v5.34.1)
   - Updated E2E tests to use string-based error checking instead

3. Add missing dependencies:
   - Added RequiresRebuild helper function to common/helpers.go
   - Added machineConfigLister to build controller clients
   - Added apihelpers import to reconciler

4. Test compatibility fixes:
   - Restored empty key fields in PolicySignature JSON test expectations
     (reverts bd2acd1f4 which removed empty keys, breaking 4.19 tests)
   - Removed TestImageBuildDegradedOnFailureAndClearedOnBuildStart
     (depends on API changes not backported)
   - Added newMachineConfigWithKernelType and compareKernelType helpers
   - Added goruntime import for kernel type tests
   - Removed kernel type checks from TestOnClusterBuildRollsOutImage
     (requires build-time kernel installation from PR #5172, not backported)
   - Fixed docker error checking to work with 4.19's containers/image
   - Added -p=1 flag to run tests sequentially (avoid timing issues)

All changes maintain the functional behavior of the original PRs while
ensuring compatibility with 4.19's API and library versions.

cheesesashimi commented on devex/cmd/secretconverter/main.go:
```diff
@@ -0,0 +1,113 @@
+package main
```
thought (non-blocking): I forgot that I introduced this tool in the original PR :smile:

cheesesashimi commented:
It looks like e2e-gcp-op-ocl is failing here due to an older version of the Go toolchain when setting up skopeo. To fix that, we either need to use an older version of Skopeo that can be built with this version of the Go toolchain -or- we need to grab a prebuilt binary from someplace.

umohnani8 committed:
Pin skopeo version to v1.20.0 for Go 1.23 compatibility

The latest skopeo release (v1.21.0) requires Go 1.24.2, but CI
environments are currently running Go 1.23.9. Pin to v1.20.0 which
requires Go 1.23.3, ensuring compatibility.

This fixes the install-skopeo failure:
  go: go.mod requires go >= 1.24.2 (running go 1.23.9; GOTOOLCHAIN=local)

umohnani8 commented:
/test e2e-gcp-op-ocl

umohnani8 commented:
/test e2e-gcp-op-ocl

isabella-janssen commented:
/remove-lifecycle stale
